Twig/Extension/SonataAdminExtension.php

```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
     }
+
+    /**
+     * Encodes the information needed by Javascript about a form widget to JSON,
+     * to be used in the association handling client-side code.
+     *
+     * @param array $context The current Twig context.
+     *
+     * @return string
+     */
+    public function renderFieldDescriptionToJSON($context)
+    {
+        /** @var AppVariable $app */
+        $app = $context['app'];
+        $request = $app->getRequest();
+        $sonataAdmin = $context['sonata_admin'];
+        /** @var Admin $admin */
+        $admin = $sonataAdmin['admin'];
+        /** @var FieldDescriptionInterface $fieldDescription */
+        $fieldDescription = $sonataAdmin['field_description'];
+        $root = $admin->getRoot();
+        $associationAdmin = $fieldDescription->getAssociationAdmin();
+        $router = $admin->getRouteGenerator();
+
+        $data = array(
+            'id'               => $context['id'],
+            'admin'            => $root->getCode(),
+            'associationAdmin' => $associationAdmin->getCode(),
+            'formType'         => $fieldDescription->getType(),
+            'editMode'         => $sonataAdmin['edit'],
+            'label'            => $associationAdmin->trans($associationAdmin->getLabel(), array(), $associationAdmin->getTranslationDomain()),
+            'routes'           => array(
+                'shortObjectDescription' => $router->generate('sonata_admin_short_object_information', array(
+                    'objectId'       => '__OBJECT_ID__',
+                    'uniqid'         => $associationAdmin->getUniqid(),
+                    'code'           => $associationAdmin->getCode(),
+                    'linkParameters' => $fieldDescription->getOption('link_parameters'),
+                )),
+                'retrieveFormElement' => $router->generate('sonata_admin_retrieve_form_element', array(
+                    'code'      => $root->getCode(),
+                    'elementId' => $context['id'],
+                    'objectId'  => $root->id($root->getSubject()),
+                    'uniqid'    => $root->getUniqid(),
+                    'subclass'  => $admin->getActiveSubclassCode(),
+                )),
+                'appendFormElement' => $router->generate(
+                    'sonata_admin_append_form_element',
+                    array_merge(array(
+                        'code'      => $root->getCode(),
+                        'elementId' => $context['id'],
+                        'objectId'  => $root->id($root->getSubject()),
+                        'uniqid'    => $root->getUniqid(),
+                        'subclass'  => $request->query->get('subclass'),
+                    ), $fieldDescription->getOption('link_parameters', array()))
+                ),
+            ),
+        );
+
+        return json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
+    }
+
+    public function renderAutocompleteOptions($context)
```

Comment:
A lot of this method will be removed once we update to select2 v4, this it allows passing options via html data attributes. So a user can configure the plugin directly using the 'attr' form option.